File: spring-security-in-action/ssia-ch07-ex3/src/main/java/org/example/ssiach07ex3/controller/TestController.java
package org.example.ssiach07ex3.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class TestController {

    @PostMapping("/a")
    public String postEndpointA() {
        return "Works!";
    }

    @GetMapping("/a")
    public String getEndPointA() {
        return "Works!";
    }

    @GetMapping("/a/b")
    public String postEndpointB() {
        return "Works!";
    }

    @GetMapping("/a/b/c")
    public String getEndPointC() {
        return "Works!";
    }

    @GetMapping("/product/{code}")
    public String postEndpointD(@PathVariable String code) {
        return code;
    }
}

File: spring-security-in-action/ssia-ch5/src/main/kotlin/org/example/ssiach5/controller/TestController.kt
package org.example.ssiach5.controller

import org.springframework.scheduling.annotation.Async
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Future

@RestController
class TestController {

    @GetMapping("/hello")
    fun hello(a: Authentication): String {
        val context = SecurityContextHolder.getContext()
        val authentication = context.authentication
        return "hello ${a?.name}!"
    }

    @GetMapping("/bye")
    @Async
    fun bye(a: Authentication): Future<String> {
        val context = SecurityContextHolder.getContext()
        val authentication = context.authentication
        return CompletableFuture.completedFuture("bye ${a?.name}!")
    }
}